solr/core/src/java/org/apache/solr/servlet/BucketedQueryRateLimiter.java

```diff
@@ -0,0 +1,234 @@
+    for (Bucket bucket : buckets) {
+      if (bucket.test(request)) {
+        bucket.tries.increment();
+        if (bucket.guaranteedSlotsPool.tryAcquire(
+            bucket.bucketCfg.slotAcquisitionTimeoutInMS, TimeUnit.MILLISECONDS)) {
+          bucket.success.increment();
+          return bucket.guaranteedSlotMetadata;
+        } else {
+          bucket.fails.increment();
+          // could not acquire  a slot
+          return RequestRateLimiter.nullSlotMetadata;
+        }
+      }
+    }
+    // no bucket matches
+    return null;
+  }
+
+  private SolrMetricsContext metrics;
+  @Override
+  public void initializeMetrics(SolrMetricsContext parentContext, String scope) {
+    metrics = parentContext.getChildContext(this);
+    MetricsMap metricsMap = new MetricsMap(ew -> buckets.forEach(bucket -> ew.putNoEx(bucket.bucketCfg.name, bucket.getMetrics())));
+    metrics.gauge(
+            metricsMap, true, scope, null, SolrInfoBean.Category.CONTAINER.toString());
+
+  }
+
+  @Override
+  public SolrMetricsContext getSolrMetricsContext() {
+    return metrics;
+  }
+
+  static class Bucket {
+    //for debugging purposes
+    private final String cfg;
+    private RateLimiterPayload.ReadBucketConfig bucketCfg;
+    private final Semaphore guaranteedSlotsPool;
+    private final SlotMetadata guaranteedSlotMetadata;
+    private final List<Condition> conditions = new ArrayList<>();
+
+    public LongAdder tries = new LongAdder();
+    public LongAdder success =new LongAdder();
+    public LongAdder fails=new LongAdder();
+    public com.codahale.metrics.Timer tryWait = new com.codahale.metrics.Timer();
+
+    private Bucket fallback;
+
+    public boolean test(RequestWrapper req) {
+      boolean isPass = true;
+      for (Condition condition : conditions) {
+        if (!condition.test(req)) isPass = false;
+      }
+      return isPass;
+    }
+
+    public MapWriter getMetrics() {
+      return ew -> {
+        ew.put("queueLength", guaranteedSlotsPool.getQueueLength());
+        ew.put("available", guaranteedSlotsPool.availablePermits());
+        ew.put("tries", tries.longValue());
+        ew.put("success", success.longValue());
+        ew.put("fails", fails.longValue());
+        ew.put("tryWaitAverage", tryWait.getMeanRate());
+      };
+    }
+
+    public Bucket(RateLimiterPayload.ReadBucketConfig bucketCfg) {
+      this.bucketCfg = bucketCfg;
+      cfg = bucketCfg.jsonStr();
+      this.guaranteedSlotsPool = new Semaphore(bucketCfg.allowedRequests);
+      this.guaranteedSlotMetadata = new SlotMetadata(guaranteedSlotsPool);
+    }
+
+    @Override
+    public String toString() {
+      return cfg;
+    }
+  }
+
+  static {
+    conditionImpls.put(ReqHeaderCondition.ID, ReqHeaderCondition.class);
+    conditionImpls.put(QueryParamCondition.ID, QueryParamCondition.class);
+  }
+
+  @SuppressWarnings("unchecked")
+  public BucketedQueryRateLimiter(RateLimiterConfig rateLimiterConfig) {
+    super(rateLimiterConfig);
```

Comment:
Perhaps we can consider enhancing the RequestRateLimiter logic in the manager at [here](https://github.com/cowpaths/fullstory-solr/pull/168/files#diff-c1b0966bd9ac92bcd42b3d72a45db56b32470b31fde317777e63e825a99bbfb4R172), instead of introducing `BucketQueryRateLimiter`?  It could avoid some code duplication, and unused `Semaphore` in parent class `QueryRateLimiter`.

It appears that we want the same ability as the existing `QueryRateLimiter` in terms of throttling (or even borrowing as @hiteshk25 mentioned in another comment), so it might be handy to just re-use `QueryRateLimiter/RequestRateLimiter` to handle bucketed limiter as well? 

The idea is during the [lookup](https://github.com/cowpaths/fullstory-solr/pull/168/files#diff-c1b0966bd9ac92bcd42b3d72a45db56b32470b31fde317777e63e825a99bbfb4R172), we could use a `Matcher` to return the correct Rate limiter. For example for non bucket configuration, it will just be the existing code (using RequestType for simple lookup from the map), but for bucket configuration, it will use iterate and check for match similar to [here](https://github.com/cowpaths/fullstory-solr/pull/168/files#diff-6c506b229d17e6251ab097b4c33063b595444f8f8a9457ea272a1097bfdfdb52R34-R35)? :)